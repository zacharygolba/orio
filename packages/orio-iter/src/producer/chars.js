// @flow

import * as result from 'orio-result'
import { AsIterator, ToString } from 'orio-traits'
import type { Producer } from 'orio-types'

@ToString
@AsIterator
export default class Chars implements Producer<*> {
  /*:: @@iterator: () => Iterator<string> */
  done: boolean
  source: Iterator<string>

  constructor(source: string) {
    this.done = false
    // $FlowFixMe
    this.source = source[Symbol.iterator]()
  }

  drop(): void {
    this.done = true
  }

  next(): IteratorResult<string, void> {
    if (this.done) {
      return result.done()
    }

    return this.source.next()
  }
}
