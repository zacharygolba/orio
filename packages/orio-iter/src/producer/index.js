// @flow

import { intoIterator } from 'orio-utils'
import type { Producer, Source } from 'orio-types'

import Chars from './chars'
import Empty from './empty'
import Indexed from './indexed'
import Once from './once'
import Unbound from './unbound'

export * from './range'
export { default as Repeat } from './repeat'
export { Chars, Empty, Indexed, Once, Unbound }

export function createProducer<T>(source: Source<T>): Producer<T> {
  if (source == null) {
    return new Empty()
  }

  if (Array.isArray(source)) {
    return new Indexed(source)
  }

  {
    const iterator = intoIterator(source)

    if (iterator != null) {
      return new Unbound(iterator)
    }
  }

  if (typeof source === 'string') {
    return new Chars(source)
  }

  return new Once((source: any))
}
