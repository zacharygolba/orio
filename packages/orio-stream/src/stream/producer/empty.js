// @flow

import * as result from 'orio-result'
import { AsAsyncIterator, ToString } from 'orio-traits'
import type { AsyncProducer, AsyncIteratorResult } from 'orio-types'

@ToString
@AsAsyncIterator
export default class Empty implements AsyncProducer<*> {
  /*:: @@asyncIterator: () => $AsyncIterator<*, void, void> */
  value: void

  drop(): void {
    this.value = undefined
  }

  async next(): AsyncIteratorResult<*, void> {
    return result.done(this.value)
  }
}
