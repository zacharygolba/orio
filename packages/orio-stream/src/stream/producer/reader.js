// @flow

import { AsAsyncIterator, ToString } from 'orio-traits'
import type {
  AsyncProducer,
  AsyncIteratorResult,
  ReadableSource,
} from 'orio-types'

@ToString
@AsAsyncIterator
export default class Reader implements AsyncProducer<*> {
  /*:: @@asyncIterator: () => $AsyncIterator<*, void, void> */
  source: ReadableSource<*>

  constructor(source: ReadableSource<*>) {
    this.source = source
  }

  drop(): void {
    this.source.cancel()
  }

  next(): AsyncIteratorResult<*, void> {
    return this.source.read()
  }
}
