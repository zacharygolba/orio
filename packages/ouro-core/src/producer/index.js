// @flow

import type { Producer, Source } from '../types'

import Chars from './chars'
import Empty from './empty'
import Indexed from './indexed'
import Once from './once'
import Unbound from './unbound'

export * from './range'
export { default as Chars } from './chars'
export { default as Empty } from './empty'
export { default as Indexed } from './indexed'
export { default as Once } from './once'
export { default as Repeat } from './repeat'
export { default as Unbound } from './unbound'

export function createProducer<+T>(source: Source<T>): Producer<T> {
  if (source == null) {
    return new Empty()
  }

  if (Array.isArray(source)) {
    return new Indexed(source)
  }

  if (typeof source === 'string') {
    return new Chars(source)
  }

  if (typeof source === 'object') {
    const iterator = source[Symbol.iterator]

    if (typeof iterator === 'function') {
      return new Unbound(iterator.call(source))
    }
  }

  return new Once((source: any))
}
