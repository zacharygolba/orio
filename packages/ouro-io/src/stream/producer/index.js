// @flow

import { intoAsyncIterator, intoIterator } from 'ouro-utils'
import type { AsyncProducer, AsyncSource } from 'ouro-types'

import * as nodeStream from '../../shims/node'
import * as webStream from '../../shims/web'
import Stream from '../stream'

import Chars from './chars'
import Empty from './empty'
import Indexed from './indexed'
import Once from './once'
import Reader from './reader'
import Task from './task'
import Unbound from './unbound'

export { Chars, Empty, Indexed, Once, Reader, Task, Unbound }

export function createProducer<T>(source: AsyncSource<T>): AsyncProducer<T> {
  if (source == null) {
    return new Empty()
  }

  if (source instanceof Promise) {
    return new Task(source)
  }

  if (source instanceof Stream) {
    return source.producer
  }

  if (webStream.isReadable(source)) {
    return new Reader(webStream.getReader(source))
  }

  if (nodeStream.isReadable(source)) {
    return new Reader(nodeStream.wrapReadable(source))
  }

  {
    const iterator = intoAsyncIterator(source)

    if (iterator != null) {
      return new Unbound(iterator)
    }
  }

  if (Array.isArray(source)) {
    return new Indexed(source)
  }

  if (typeof source === 'string') {
    return new Chars(source)
  }

  {
    const iterator = intoIterator(source)

    if (iterator != null) {
      return new Unbound(iterator)
    }
  }

  return new Once((source: any))
}
