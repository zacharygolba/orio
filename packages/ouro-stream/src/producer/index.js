// @flow

import { intoIterator } from 'ouro-utils'
import type { AsyncProducer, AsyncSource } from 'ouro-types'

import Chars from './chars'
import Empty from './empty'
import Indexed from './indexed'
import Once from './once'
import Task from './task'
import Unbound from './unbound'

export { default as Chars } from './chars'
export { default as Empty } from './empty'
export { default as Indexed } from './indexed'
export { default as Once } from './once'
export { default as Task } from './task'
export { default as Unbound } from './unbound'

export function createProducer<+T>(source: AsyncSource<T>): AsyncProducer<T> {
  if (source == null) {
    return new Empty()
  }

  if (source instanceof Promise) {
    return new Task(source)
  }

  if (Array.isArray(source)) {
    return new Indexed(source)
  }

  if (typeof source === 'string') {
    return new Chars(source)
  }

  {
    const iterator = intoIterator(source)

    if (iterator != null) {
      return new Unbound(iterator)
    }
  }

  return new Once((source: any))
}
