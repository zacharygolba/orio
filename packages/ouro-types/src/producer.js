// @flow

import type { AsyncIterator } from './iterator'

export interface Drop { drop(): void }
export interface Producer<+T> extends Drop, Iterable<T>, Iterator<T> {}
export interface AsyncProducer<+T>
  extends AsyncIterable<T>, AsyncIterator<T>, Drop {}
