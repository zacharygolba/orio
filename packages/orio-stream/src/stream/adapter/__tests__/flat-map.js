// @flow

import FlatMap from '../flat-map'
import { createProducer } from '../../producer'

let subj

const flatMap = producer => {
  const adapter = new FlatMap(producer, item => Promise.resolve([item, item]))

  jest.spyOn(adapter.parent, 'drop')
  return adapter
}

beforeEach(() => {
  const producer = createProducer(Promise.resolve([1, 2, 3]))
  subj = flatMap(producer)
})

afterEach(() => {
  jest.resetAllMocks()
})

test('#drop()', () => {
  expect(subj.drop()).toBeUndefined()
  expect(subj.parent.drop).toHaveBeenCalled()
})

test('#next()', async () => {
  expect(await subj.next()).toMatchSnapshot()
  expect(await subj.next()).toMatchSnapshot()
  expect(await subj.next()).toMatchSnapshot()
  expect(await subj.next()).toMatchSnapshot()
  expect(await subj.next()).toMatchSnapshot()
  expect(await subj.next()).toMatchSnapshot()
  expect(await subj.next()).toMatchSnapshot()
})
