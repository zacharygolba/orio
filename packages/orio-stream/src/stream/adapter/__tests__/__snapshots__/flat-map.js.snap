// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#next() 1`] = `
Object {
  "done": false,
  "value": 1,
}
`;

exports[`#next() 2`] = `
Object {
  "done": false,
  "value": 1,
}
`;

exports[`#next() 3`] = `
Object {
  "done": false,
  "value": 2,
}
`;

exports[`#next() 4`] = `
Object {
  "done": false,
  "value": 2,
}
`;

exports[`#next() 5`] = `
Object {
  "done": false,
  "value": 3,
}
`;

exports[`#next() 6`] = `
Object {
  "done": false,
  "value": 3,
}
`;

exports[`#next() 7`] = `
Object {
  "done": true,
  "value": undefined,
}
`;
